src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+                        convertingNodes.Add(currentForEachStatement);
+                        current = currentForEachStatement.Statement;
+                        break;
+
+                    case SyntaxKind.IfStatement:
+                        var ifStatement = (IfStatementSyntax)current;
+                        if (ifStatement.Else == null)
+                        {
+                            convertingNodes.Add(ifStatement);
+                            current = ifStatement.Statement;
+                            break;
+                        }
+                        else
+                        {
+                            statementsCannotBeConverted = new[] { current };
+                            break;
+                        }
+
+                    case SyntaxKind.LocalDeclarationStatement:
+                        // This is a situation with "var a = something;" s the most internal statements inside the loop.
+                        ProcessLocalDeclarationStatement((LocalDeclarationStatementSyntax)current);
+                        statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        break;
+
+                    case SyntaxKind.EmptyStatement:
+                        statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
+                        break;
+
+                    default:
+                        statementsCannotBeConverted = new[] { current };
+                        break;
+                }
+            }
+
+            return (convertingNodes.ToImmutableArray(), identifiers.ToImmutableArray(), statementsCannotBeConverted.ToImmutableArray());
+        }
+
+        protected override bool TryBuildSpecificConverter(
+            ForEachStatementSyntax forEachStatement,
+            SemanticModel semanticModel,
+            ImmutableArray<SyntaxNode> convertingNodes,
+            StatementSyntax statementCannotBeConverted,
+            CancellationToken cancellationToken,
+            out IConverter converter)
+        {
+            switch (statementCannotBeConverted.Kind())
+            {
+                case SyntaxKind.ExpressionStatement:
+                    var expression = ((ExpressionStatementSyntax)statementCannotBeConverted).Expression;
+                    switch (expression.Kind())
+                    {
+                        case SyntaxKind.PostIncrementExpression:
+                            // No matter what can be used as the last select statement for the case of Count. We use SyntaxFactory.IdentifierName(forEachStatement.Identifier).
+                            converter = new ConvertToCountSyntaxEditor(
+                                forEachStatement,
+                                convertingNodes,
+                                SyntaxFactory.IdentifierName(forEachStatement.Identifier),
+                                ((PostfixUnaryExpressionSyntax)expression).Operand);
+                            return true;
```

Comment:
all these sections would benefit from explanations specifying what they're matching against, and what they're generating.